README.md
# Kafka header counter for prometheus

This simple script developed in order to count a specefic header values from
kafka messages. For example, we are going to count the messages fetched by
individual user IDs by twitter crawler.

## Configurations

A sample config file provided in `config.example.yml`. It's **required** to
mount a config file inside the cotainer with the name `config.yml` at
`/app/config.yml` or provide the config file absoulte path with environment
variable `CONFIG_ADDRESS`.

### config example

``` yaml
## Required configs ##
## just name the service for exported descriptions
service: example_service
## counter name for exported metrics
counter_name: test_counter_name
## kafka topics to fetch messages from
topics:
  - test_topic_a
  - test_topic_b
## consumer group name of this exporter
consumer: test_consumer_group
## a list of kafka addresses with port
address:
  - kafka.example.com:9092
## header name to fetch data from it
header_fields:
  - test_header_field

## Optional configs ##
## timeouts in ms
timeout: 6000
## offset reset policy, can be earliest or latest
offset_reset: earliest
## exposed http server port for prometheus scraper
port: 80
```

## Type

RESTful API

## Source

Apache kafka

## Output

Prometheus

## Tests

Not developed yet

## Deployment Notes

### Hardware Requirements

- Memory: 100 Mi - 128 Mi
- CPU: 0.7 - 1
- Storage: None


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [v0.1.5] - 2022-02-22

### Added

- Added the [sentry](https://sentry.io) error tracking capability

### Changed

- Fixed [`requirements.txt`](requirements.txt) items versions

## [v0.1.4] - 2021-09-17

### Changed

- Multiple header fields to export instead of one
- Enhanced runtime performance

## [v0.1.3] - 2021-09-08

### Fixed

- Fixed delay between messages

## [v0.1.2] - 2021-09-08

### Added

- Added topic and service name to the metrics

### Fixed

- Added `PYTHONUNBUFFERED=1` to the `Dockerfile` to fix fetching logs in the
  Kubernetes

## [v0.1.1] - 2021-09-07

### Fixed

- Handled some exceptions

## [v0.1.0] - 2021-09-06

### Added

- Initial version


app.py
#!/usr/bin/env python

"""An exporter to fetch a fields value and count it for prometheus."""

from time import sleep

import prometheus_client as prom
import sentry_sdk

from broker import KafkaHandler
from config import Config


service = Config().service
counter_name = Config().counter_name
header_fields = Config().header_fields
counter = prom.Counter(
    "{}_{}".format(service, counter_name),
    "The {}'s {} counter".format(service, (header_fields)),
    header_fields + ["service", "topic"]
)
exposed_port = Config().port
sentry = Config().sentry

if sentry is not None:
    sentry_sdk.init(
        dsn = sentry.get("dsn"),
        release = sentry.get("release"),
        environment = sentry.get("environment"),
        send_default_pii = sentry.get("send-default-pii"),
        debug = sentry.get("debug"),
        traces_sample_rate = sentry.get("traces-sample-rate")
    )

welcome_message = \
    "Daemon started successfully on port {} ...".format(exposed_port)


if __name__ == "__main__":
    prom.start_http_server(exposed_port)
    consumer = KafkaHandler()
    print(welcome_message)
    sentry_sdk.capture_message(welcome_message)
    for message in consumer.consume_loop():
        try:
            headers = {k: v.decode('utf-8')
                       for k, v in dict(message.headers()).items()}
            headers['service'] = service
            headers['topic'] = message.topic()
        except Exception as e:
            print("ERROR: ", e)
            sentry_sdk.set_context("message-headers", message.headers())
            sentry_sdk.capture_exception(e)
        finally:
            consumer.try_commit()
            try:
                counter.labels(**headers).inc()
            except Exception as e:
                print("ERROR: Error in mapping headers to configured ones", e)
                sentry_sdk.set_context("prometheus-labels", headers)
                sentry_sdk.capture_exception(e)


broker.py
"""An kafka consumer handler wich loops on the messages and commit method."""

from confluent_kafka import Consumer
from confluent_kafka.cimpl import KafkaError, KafkaException

from config import Config


kafka = Config()
topics = kafka.topics
address = ",".join(kafka.address)


class KafkaHandler:
    def __init__(
            self,
            address=address,
            group_id=kafka.consumer,
            topics=topics):
        connection = {'bootstrap.servers': address,
                      'group.id': group_id,
                      'auto.offset.reset': kafka.offset_reset,
                      'session.timeout.ms': kafka.timeout,
                      'enable.auto.commit': False}
        self.consumer = Consumer(connection)
        self.msg_count = 0
        self.MIN_COMMIT_COUNT = 1
        self.consumer.subscribe(topics)

    def consume_loop(self):
        try:
            running = True
            while running:
                msg = self.consumer.poll(timeout=1.0)
                if msg is None:
                    continue
                if msg.error():
                    if msg.error().code() == KafkaError._PARTITION_EOF:
                        # End of partition event
                        print('%% %s [%d] reached end at offset %d\n' %
                              (msg.topic(), msg.partition(), msg.offset()))
                    elif msg.error():
                        raise KafkaException(msg.error())
                else:
                    yield msg
        finally:
            # Close down consumer to commit final offsets.
            self.consumer.close()

    def try_commit(self):
        self.msg_count += 1
        if self.msg_count % self.MIN_COMMIT_COUNT == 0:
            self.consumer.commit(asynchronous=False)
            self.msg_count = 0


config.py
from os import path, getenv
import yaml


class Config:

    """
    A simple handler for kafka configurations. The methods of this are used to
    handle config properties used to configure kafka client
    """

    def __init__(self, config_file="config.yml"):
        """
        :type topics: list
        :type config_file: str
        :arg topics:
            Containing list of the topics to be read
        :arg config_file:
            Containing the file path to config file
        """

        # The configuration is set in a yml file.
        dirname = path.dirname(path.abspath(__file__))
        absolute_path = str(dirname + "/" + config_file)
        self.config_file = getenv("CONFIG_ADDRESS", absolute_path)
        with open(self.config_file, "r") as ymlfile:
            config = yaml.safe_load(ymlfile)
        self.service = config["service"]
        self.counter_name = config["counter_name"]
        self.topics = config["topics"]
        self.consumer = config["consumer"]
        self.address = config["address"]
        self.header_fields = config["header_fields"]
        self.timeout = config.get("timeout", 6000)
        self.offset_reset = config.get("offset_reset", "earliest")
        self.port = config.get("port", 80)
        if config.get("sentry", None) is not None:
            self.sentry = config.get("sentry")
            self.sentry["dsn"] = self.sentry.get("dsn", None)
            self.sentry["release"] = self.sentry.get("release", None)
            self.sentry["environment"] = self.sentry.get(
                "environment", "production")
            self.sentry["send-default-pii"] = self.sentry.get(
                "send-default-pii", False)
            self.sentry["debug"] = self.sentry.get("debug", False)
            self.sentry["traces-sample-rate"] = self.sentry.get(
                "traces-sample-rate", 1.0)
        else:
            self.sentry = None
